### PR TITLE
Fix CSI e2e leaving pods in terminating state

### PR DESCRIPTION
**Pull request by gnufied**
This PR fixes problem with CSI tests leaving pods in terminating state when testsuite is interrupted . This happens because driver pod gets deleted before testpods have a chance to terminate.

This PR solves this problem by creating driver pods in different namespace than the test pod.

This is kinda another iteration of https://github.com/kubernetes/kubernetes/pull/89944

/sig storage

### CODE REVIEW

**Comment by gnufied** [comment]
I know tests are failing and I am working on fixing them.


/assign @pohly @msau42 @jsafrane

File: test/e2e/framework/framework.go
Comment (pohly) [inline test/e2e/framework/framework.go:87]: Why is this needed inside the framework? If the CSI driver deployment code is the only place where this is needed, then it may be better handle the second namespace there.

Comment (gnufied) [reply test/e2e/framework/framework.go:87]: Having secondary namespace being tracked here has several benefits. 
- The code in framework.go ensures any registered namespace is deleted even if somehow it is skipped from storage testsuite.
- It ensures proper reporting and dumping of events in case there is an error deleting the namespace etc.

Comment (pohly) [reply test/e2e/framework/framework.go:87]: Why can't the same be done in the storage test suite? Tests should have the same access to cleanup and logging as the framework itself.

Comment (gnufied) [reply test/e2e/framework/framework.go:87]: may be. But I do not see any inherent advantage of doing this in storage test suite vs doing it here. I also forgot to mention - if there are projects that use k8s e2e as a library then having the namespace exposed in framework.go allows them to customize it. Such as apply extra permissions etc.

Comment (pohly) [reply test/e2e/framework/framework.go:87]: I'm asking because there has been a recent push slim down the API of the framework, including the removal of code that was only used by some tests. For example, podlog was originally added to the framework "because it might be useful to others" and then moved under test/e2e/storage because in practice it was only used by storage tests.

Comment (gnufied) [reply test/e2e/framework/framework.go:87]: okay, but in this case IMO it makes sense to have this in framework.go because moving this code to storage suite will result in unnecessary duplication.

Comment (pohly) [reply test/e2e/framework/framework.go:87]: The cleanup enhancements in the framework belong there, but I don't see the need for `CreateSecondaryNamespace()` and `f.SecondaryNamespace` in the framework.

What will be duplicated when moving that into `test/e2e/storage`? `CreateSecondaryNamespace` only uses exported functionality of the framework, so it can be implemented outside of the framework without copying code.

Comment (gnufied) [reply test/e2e/framework/framework.go:87]: yes the function itself has to be copied and then following bits of code that runs in `AfterEach` of `framework.go` needs to be copied:

```
if err := f.ClientSet.CoreV1().Namespaces().Delete(context.TODO(), ns.Name, metav1.DeleteOptions{}); err != nil {
					if !apierrors.IsNotFound(err) {
						nsDeletionErrors[ns.Name] = err

						// Dump namespace if we are unable to delete the namespace and the dump was not already performed.
						if !ginkgo.CurrentGinkgoTestDescription().Failed && TestContext.DumpLogsOnFailure {
							DumpAllNamespaceInfo(f.ClientSet, ns.Name)
						}
					} else {
						Logf("Namespace %v was already deleted", ns.Name)
					}
				}
```

So as namespace info could be dumped. I have not looked very closely though there could be more stuff and that is what worries me.

Comment (pohly) [reply test/e2e/framework/framework.go:87]: Perhaps the framework should have a generic API for deleting additional namespaces in a guaranteed order and a function for creating them? That looks like a cleaner approach to me than `f.SecondaryNamespace` - what if a test needs three or more namespaces...

Comment (gnufied) [reply test/e2e/framework/framework.go:87]: hmm okay but I am unsure about how that API will look like. @pohly do you think it is a blocker issue for merging this PR?

Another thing about creating that generic API is - if it exists at framework initialization time then it has to be triggered via testsuite whereas creating secondary namespace for now is part of CSI driver. So basically, a storage test that runs with in-tree EBS and CSI gce-pd driver behave differently and framework initialization happens at the top of e2e test file.

Comment (pohly) [reply test/e2e/framework/framework.go:87]: > hmm okay but I am unsure about how that API will look like. @pohly do you think it is a blocker issue for merging this PR?

Given that moving the framework into staging is a goal for 1.19 (https://github.com/kubernetes/enhancements/issues/1746) and that this implies stabilizing the API, I'm worried about this rather ad-hoc API change. But I'll let @alejandrox1 and @oomichi decide this.

Comment (andrewsykim) [inline test/e2e/framework/framework.go:87]: Can we make `Namespace` into `Namespaces` instead?

Comment (gnufied) [reply test/e2e/framework/framework.go]: I have removed this bit entirely from the code and move namespace creation to storage utils.

Comment (gnufied) [reply test/e2e/framework/framework.go]: okay, so I moved most of namespace creation code to storage/utils.go . The namespace cleanup should still happen because framework.go tracks these namespaces in `namespacesToDelete` variable.

File: test/e2e/framework/framework.go
Comment (pohly) [inline test/e2e/framework/framework.go]: Nit: why this change? The previous line did the same thing.

File: test/e2e/framework/framework.go
Comment (pohly) [inline test/e2e/framework/framework.go]: Nit: this empty line probably could be removed.

File: test/e2e/storage/utils/create.go
Comment (pohly) [inline test/e2e/storage/utils/create.go:144]: The description of `CreateItems` must be adapted. It currently still says:
```
// Cleaning up after a test can be triggered in two ways:
// - the test invokes the returned cleanup function,
//   usually in an AfterEach
// - the test suite terminates, potentially after
//   skipping the test's AfterEach (https://github.com/onsi/ginkgo/issues/222)
```

The new behavior is that there is only one way of deleting the created objects.

**Comment by gnufied** [comment]
cc @sig-testing-approvers

**Review by pohly — COMMENTED** [review]
Some minor nits, but overall the diff now looks reasonable. Before we merge, can you squash the commits? Some commits are fixes previous commits, and in the middle we have the approach with secondary namespace which then gets removed again.

File: test/e2e/storage/utils/create.go
Comment (gnufied) [reply test/e2e/storage/utils/create.go:144]: it can still get called from `AfterSuite` hooks via  - https://github.com/kubernetes/kubernetes/pull/90773/files#diff-4d73861d7cac72f86d4c7bccdc4e594fR226 though.

File: test/e2e/framework/framework.go
Comment (gnufied) [reply test/e2e/framework/framework.go]: this code actually got reverted. You might have commented on outdated diff. Let me squash the commits, so as it goes away entirely.

File: test/e2e/framework/framework.go
Comment (gnufied) [reply test/e2e/framework/framework.go]: same as above. changes into `framework.go` are entirely removed.

File: test/e2e/framework/framework.go
Comment (gnufied) [reply test/e2e/framework/framework.go]: So I have squashed the commits but I still wanted changes into `AfterSuite` hook ordering in separate commit. This does mean some of old bits are still in that commit which gets dropped in second commit.

Comment (gnufied) [reply test/e2e/framework/framework.go]: Alright. Fixed up first commit to drop changes into framework files.

**Review by pohly — APPROVED** [review]
Thanks for your patience.

/lgtm

**Comment by pohly** [comment]
If tests are clean and there are no objections, then I'll also approve.

**Comment by pohly** [comment]
Test failures seem unrelated, so:
/approve

**Pull request merged**
